File: 2018/day02-2.hs
import System.IO

solution :: String -> String
solution content = map (\(x,_) -> x) . filter (\(a,b) -> a == b) $ zip fst snd
    where
        (fst, snd) = [(fst, snd) | fst <- lines content, snd <- lines content, (distance fst snd) == 1] !! 0

distance :: String -> String -> Int
distance "" "" = 0
distance (x:xs) (y:ys)
    | x /= y = 1 + distance xs ys
    | otherwise = distance xs ys

main :: IO ()
main = do
    handle <- openFile "day02.dat" ReadMode
    content <- hGetContents handle
    print . solution $ content

File: 2018/day02-2.ts
function solution(content: string) {
  const [fst, snd] = content.split("\n").reduce((acc, fst) => {
    if (acc) return acc;
    const similar = content
      .split("\n")
      .filter((snd) => distance(fst, snd) === 1);
    if (similar.length) return [fst, similar.at(0)!];
    return null;
  }, null as [string, string] | null)!;

  return fst
    .split("")
    .map((a, index) => (a === snd[index] ? a : ""))
    .join("");
}

function distance(fst: string, snd: string) {
  return fst
    .split("")
    .map<number>((a, index) => (a === snd[index] ? 0 : 1))
    .reduce((acc, cur) => acc + cur, 0);
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
